econml/solutions/causal_analysis/_causal_analysis.py

```diff
@@ -628,21 +658,54 @@ def fit(self, X, y, warm_start=False):
                 f"has length {len(categories)} while there are {len(categorical_inds)} categorical columns.")
 
         # check for indices over the categorical expansion bound
-        over_bound_inds = []
+        invalid_inds = getattr(self, 'untrained_feature_indices_', [])
+
+        # assume we'll be able to train former failures this time; we'll add them back if not
+        invalid_inds = [(ind, reason) for (ind, reason) in invalid_inds if ind not in new_inds]
+
+        min_counts = {}
         for ind in new_inds:
-            n_cats = len(np.unique(_safe_indexing(X, ind, axis=1)))
-            if ind in categorical_inds and n_cats > self.upper_bound_on_cat_expansion:
-                warnings.warn(f"Column {ind} has more than {self.upper_bound_on_cat_expansion} values "
-                              "so no heterogeneity model will be fit for it; increase 'upper_bound_on_cat_expansion' "
-                              "to change this behavior.")
-                # can't remove in place while iterating over new_inds, so store in separate list
-                over_bound_inds.append(ind)
-        for ind in over_bound_inds:
+
+            if ind in categorical_inds:
+                cats, counts = np.unique(_safe_indexing(X, ind, axis=1), return_counts=True)
+                min_ind = np.argmin(counts)
+                if len(cats) > self.upper_bound_on_cat_expansion:
+                    warnings.warn(f"Column {ind} has more than {self.upper_bound_on_cat_expansion} values "
+                                  "so no heterogeneity model will be fit for it; increase "
+                                  "'upper_bound_on_cat_expansion' to change this behavior.")
+                    # can't remove in place while iterating over new_inds, so store in separate list
+                    invalid_inds.append((ind, 'upper_bound_on_cat_expansion'))
+
+                elif counts[min_ind] < _CAT_LIMIT:
+                    if self.skip_cat_limit_checks and counts[min_ind] >= 2:
+                        # train the model, but warn
+                        warnings.warn(f"Column {ind}'s value {cats[min_ind]} has only {counts[min_ind]} instances in "
+                                      f"the training dataset, which is less than {_CAT_LIMIT}. A model will be fit "
+                                      "because 'skip_cat_limit_checks' is True, but this model may not be robust.")
+                        min_counts[ind] = counts[min_ind]
+                    elif counts[min_ind] < 2:
```

Comment:
two questions here:
1. Is 2 enough? I heard this morning you said we might have tuning for CausalForestDML will split again internally?
2. the warning below should say the minimum instances is less than 2 explicitly? No matter check cat or not, this can't be trained.

Reply:
Both of these issues have been addressed in the latest commit.